### DISCUSSION TITLE
disable multiline

### DISCUSSION THREAD

**Discussion by ArianaAsl** [akinomyoga/ble.sh#199] [Q&A] [answered]
Hello 
I've read the wiki and the Q&A and the manual but can't find a way to disable the multi line option completely.
I tried increasing the threshold but that did nothing.
I really like ble.sh and would like to keep using it without the multi line editing mode.
any help is appreciated 
thanks.

**Comment by akinomyoga** [comment, marked as answer]
Thank you for your feedback. I think I need to ask about what you exactly mean by *disabling the multiline editing mode completely*. The multiline editing support affects a wide range of behaviors of `ble.sh`, and disabling some of them doesn't seem to make a sense. For example, when the multiline editing mode is completely disabled, what to do with an existing multiline command-history entry?

Would you like to delete all the newlines in the command line even when you perform the quoted insert or the terminal's bracketed paste? Would you like to also delete all the existing history entries that contain newlines or like to handle them in another way? Or does it just mean that you would like to still run the command by <kbd>RET</kbd> or <kbd>C-m</kbd> within the multiline editing mode? Or would you still like to edit the command line containing newlines but like to display the newline character as a control character `^J` in the terminal? I'd like to hear from you what you exactly mean by disabling the multiline editing mode or which specific behaviors you would like to turn off.

**Reply by ArianaAsl** [reply]
I apologize for the lack of detail in my initial question and thank you for getting back to me so quickly. To simplify my need, all i need from ble.sh is the completion and syntax highlighting and maybe vi mode and all of them work great. However I have no use for a multiline mode since I don't enter multiline commands into my shell. All I do at the shell is pretty much single line. I'm sure its a feature that a lot of people like and judging from having read many of the codes that come with ble.sh it took a lot of effort to implement, yet it is not suitable for my use-case. Although I should add that if multiline mode allowed me to press enter to run my command instead of Ctrl+j. that would also be ok for me. 

For example: 
I have been designing a prompt in bash and pasting the PS1='\[\e[0;38;5;208m\]┬─[\[\e[0;38;5;208m\]\w\n\[\e[0;38;5;208m\]╰─>\[\e[0m\]' after [export] got me immediately into multiline which was uncalled for.

Or using my single letter aliases like r for ranger or v for vim also gets me into multiline mode sometimes.

**Reply by akinomyoga** [reply]
Thank you for your response!

> Although I should add that if multiline mode allowed me to press enter to run my command instead of Ctrl+j. that would also be ok for me.

That should be achieved by the following settings:

```bash
# for vi mode
ble-bind -m vi_imap -f 'C-m' 'accept-line'
ble-bind -m vi_imap -f 'RET' 'accept-line'
ble-bind -m vi_nmap -f 'C-m' 'accept-line'
ble-bind -m vi_nmap -f 'RET' 'accept-line'

# for emacs mode
ble-bind -m emacs -f 'C-m' 'accept-line'
ble-bind -m emacs -f 'RET' 'accept-line'
```

> pasting the PS1='[\e[0;38;5;208m]┬─[[\e[0;38;5;208m]\w\n[\e[0;38;5;208m]╰─>[\e[0m]' after [export] got me immediately into multiline which was uncalled for.

By default, `ble.sh` does not execute the commands by the newlines coming from the pasted text. Instead, newlines contained in pasted texts are treated as a part of the command line. This is intentional behavior in order not to execute random texts mistakenly pasted into the terminal. `ble.sh` uses two ways to detect the pasted texts:

1. One is the detection by the speed of input. which can be disabled by binding `accept-line` as described above or by setting `bleopt accept_line_threashold=-1`.
2. The other is the detection by the terminal support for the bracketed paste mode (aka DEC Mode 2004). Currently, `ble.sh` does not have an option to disable the use of the terminal's bracketed paste mode, i.e., it is currently always turned on as far as your terminal supports it.

I guess the behavior you faced with by pasting `PS1` has been detected through the second mechanism. I think I can implement a new option to disable the second detection, but before implementing it I would like to confirm if that is what you want. The following settings have the same effect as the proposed new option. Could you try the following settings to see if the behavior becomes what you want?

```bash
# blerc
function ble/term/bracketed-paste-mode/enter { :; }
function ble/term/bracketed-paste-mode/leave { :; }
```

> Or using my single letter aliases like r for ranger or v for vim also gets me into multiline mode sometimes.

I currently do not have an idea why they trigger the multiline mode. Maybe this is caused by something that I missed. I would appreciate it if you could describe the situation in more detail.

**Reply by ArianaAsl** [reply]
Thanks! I added these lines to my .blerc and its working perfectly fine now as far as I can tell by trying to recreate the problem. I will do further tests and report to you. Thank you for your time. Regards.
